lib/stats/stats-query.c

```diff
@@ -117,62 +96,74 @@ _is_single_match(const gchar *key_str)
   return !strpbrk(key_str, "*?");
 }
 
-static GList *
-_query_counter_hash(const gchar *key_str)
+static void
+_process_counter_if_matching(StatsCluster *sc, gint type, StatsCounterItem *counter,
+                             gpointer user_data /*, gboolean* cancelled*/)
 {
-  GPatternSpec *pattern = g_pattern_spec_new(key_str);
-  GList *counters = NULL;
-  gpointer key, value;
-  GHashTableIter iter;
-  gboolean single_match;
-
-  _update_index();
-  single_match = _is_single_match(key_str);
-
-  g_mutex_lock(&stats_query_mutex);
-  g_hash_table_iter_init(&iter, counter_index);
-  while (g_hash_table_iter_next(&iter, &key, &value))
+  gpointer *args = (gpointer *) user_data;
+  int arg_ndx = 0;
+  GPatternSpec *pattern = (GPatternSpec *) args[arg_ndx++];
+  gboolean single_match = *(gboolean const *) args[arg_ndx++];
+  ProcessCounterCb process_func = (ProcessCounterCb) args[arg_ndx++];
+  gpointer process_func_user_data = args[arg_ndx++];
+  StatsFormatCb format_cb = (StatsFormatCb) args[arg_ndx++];
+  gpointer format_cb_user_data = args[arg_ndx++];
+  gboolean must_reset = *(gboolean const *) args[arg_ndx++];
+  gboolean *found = (gboolean *) args[arg_ndx++];
+
+  if (_update_counter_index(sc, type, counter))
     {
-      if (_is_pattern_matches_key(pattern, key))
+      if (_is_pattern_matches_key(pattern, counter->name))
         {
-          StatsCounterItem *counter = (StatsCounterItem *) value;
-          counters = g_list_prepend(counters, counter);
+          *found = TRUE;
+
+          process_func(counter, process_func_user_data, format_cb, format_cb_user_data);
+          if (must_reset)
+            stats_control_reset_counter_if_needed(sc, counter);
 
           if (single_match)
-            break;
+            {
+              /* DISCUSS: cancelled parameter is not implemented yet in the counter enumerators, only in the cluster ones, and
+                          - not stopping here now does not make a performace difference as the original impl always interated over all the counter items in all the clusters
+                          - stopping here now will not update the indexed state (indexed_mask) of all of the clusters leaving them in an incosistent state (that might be acceptable
+                            as currently only this query module uses that state)
```

Comment:
I think we need to drop the indexed_mask entirely. that will simplify this code even more. This code keeps locks for an extended amount of time, which is almost certainly a cause for stalls in processing of the production traffic... When a new connection comes in, it would allocate counters, which is not possible until the stats_lock() is kept. Which it is for any kind of stats-query, which does multiple passes over _every_ counter registered.

Reply:
I remove the indexed_mask and the related indexed state management code parts, but that will not change this at all.
During the enumeration and the item manipultions the stats_lock() still must be held.
BTW, there is no multiple passes here anymore, we have only one single pass now (that of course still can hold the lock for a while if there are 100 thousands of stats items)

Reply:
This comment was related to the missing possibility, which the original code has, braking the enumeration at any given point, like if the query has a single match only